File: generate_data.py

import json
import random
from datetime import datetime, timedelta

def random_range(min_val, max_val):
    return random.uniform(min_val, max_val)

def random_int(min_val, max_val):
    return random.randint(min_val, max_val)

def format_date(dt):
    return dt.strftime('%Y.%m.%d %H:%M:%S')

trades = []
num_trades = 75
target_total_profit = 350.0
current_total_profit = 0.0

# Generate 75 trades
for i in range(num_trades):
    is_buy = random.choice([True, False])
    deal_type = 'buy' if is_buy else 'sell'
    volume = 0.01
    
    # Price between 3850 and 3950
    open_price = round(random_range(3850.0, 3950.0), 2)
    
    # Profit between 2.00 and 8.00
    profit = round(random_range(2.00, 8.00), 2)
    
    close_price = 0.0
    if deal_type == 'buy':
        close_price = round(open_price + profit, 2)
    else:
        close_price = round(open_price - profit, 2)
    
    # Date generation
    day = random_int(1, 31)
    hour = random_int(0, 23)
    minute = random_int(0, 59)
    second = random_int(0, 59)
    
    close_time_date = datetime(2025, 10, day, hour, minute, second)
    
    # Open time 5 mins to 4 hours before
    duration_minutes = random_int(5, 240)
    open_time_date = close_time_date - timedelta(minutes=duration_minutes)

    trades.append({
        'type': deal_type,
        'volume': volume,
        'symbol': 'XAUUSD',
        'openPrice': open_price,
        'closePrice': close_price,
        'openTimeDate': open_time_date,
        'closeTimeDate': close_time_date,
        'profit': profit,
        'sl': 0,
        'tp': 0,
        'commission': 0,
        'swap': 0
    })
    
    current_total_profit += profit

# Adjust profits to meet target ~350
if current_total_profit != 0:
    scale_factor = target_total_profit / current_total_profit
else:
    scale_factor = 1.0

for t in trades:
    new_profit = round(t['profit'] * scale_factor, 2)
    # Ensure range 2.00 - 8.00
    if new_profit < 2.00:
        new_profit = 2.00
    elif new_profit > 8.00:
        new_profit = 8.00
    
    t['profit'] = new_profit
    
    # Recalculate Close Price
    if t['type'] == 'buy':
        t['closePrice'] = round(t['openPrice'] + t['profit'], 2)
    else:
        t['closePrice'] = round(t['openPrice'] - t['profit'], 2)

# Add Withdrawal
withdrawal_date = datetime(2025, 10, 24, 10, 5, 42)
trades.append({
    'type': 'balance',
    'volume': 0,
    'symbol': '',
    'openPrice': 0,
    'closePrice': 0,
    'openTimeDate': withdrawal_date,
    'closeTimeDate': withdrawal_date,
    'profit': -150.0,
    'sl': 0,
    'tp': 0,
    'commission': 0,
    'swap': 0
})

# Sort by close time
trades.sort(key=lambda x: x['closeTimeDate'])

# Assign IDs
start_id = 514343326
final_trades = []

for t in trades:
    deal = {
        'id': str(start_id),
        'symbol': t['symbol'],
        'type': t['type'],
        'volume': t['volume'],
        'openTime': format_date(t['openTimeDate']),
        'closeTime': format_date(t['closeTimeDate']),
        'openPrice': t['openPrice'],
        'closePrice': t['closePrice'],
        'sl': t['sl'],
        'tp': t['tp'],
        'commission': t['commission'],
        'swap': t['swap'],
        'profit': t['profit']
    }
    
    if t['type'] == 'balance':
        deal['symbol'] = ''
        deal['openPrice'] = 0
        deal['closePrice'] = 0
        deal['openTime'] = deal['closeTime']
        
    final_trades.append(deal)
    start_id += 1

with open('result.json', 'w') as f:
    json.dump(final_trades, f, indent=2)

print("Done")

File: generate_data.js

const fs = require('fs');

try {
    function randomRange(min, max) {
        return Math.random() * (max - min) + min;
    }

    function randomInt(min, max) {
        return Math.floor(Math.random() * (max - min + 1)) + min;
    }

    function formatDate(date) {
        const yyyy = date.getFullYear();
        const mm = String(date.getMonth() + 1).padStart(2, '0');
        const dd = String(date.getDate()).padStart(2, '0');
        const hh = String(date.getHours()).padStart(2, '0');
        const min = String(date.getMinutes()).padStart(2, '0');
        const ss = String(date.getSeconds()).padStart(2, '0');
        return `${yyyy}.${mm}.${dd} ${hh}:${min}:${ss}`;
    }

    const trades = [];
    const numTrades = 75;
    const targetTotalProfit = 350;
    let currentTotalProfit = 0;

    // Generate 75 trades
    for (let i = 0; i < numTrades; i++) {
        const isBuy = Math.random() > 0.5;
        const type = isBuy ? 'buy' : 'sell';
        const volume = 0.01;
        
        // Price between 3850 and 3950
        const openPrice = parseFloat(randomRange(3850, 3950).toFixed(2));
        
        // Profit between 2.00 and 8.00
        let profit = parseFloat(randomRange(2.00, 8.00).toFixed(2));
        
        let closePrice;
        if (type === 'buy') {
            closePrice = openPrice + profit;
        } else {
            closePrice = openPrice - profit;
        }
        closePrice = parseFloat(closePrice.toFixed(2));
        
        // Date generation
        // October has 31 days
        const day = randomInt(1, 31);
        const hour = randomInt(0, 23);
        const minute = randomInt(0, 59);
        const second = randomInt(0, 59);
        
        const closeTimeDate = new Date(2025, 9, day, hour, minute, second); // Month is 0-indexed (9 = Oct)
        
        // Open time 5 mins to 4 hours before
        const durationMs = randomInt(5 * 60 * 1000, 4 * 60 * 60 * 1000);
        const openTimeDate = new Date(closeTimeDate.getTime() - durationMs);

        trades.push({
            type: type,
            volume: volume,
            symbol: 'XAUUSD',
            openPrice: openPrice,
            closePrice: closePrice,
            openTimeDate: openTimeDate,
            closeTimeDate: closeTimeDate,
            profit: profit,
            sl: 0,
            tp: 0,
            commission: 0,
            swap: 0
        });
        
        currentTotalProfit += profit;
    }

    const scaleFactor = targetTotalProfit / currentTotalProfit;
    trades.forEach(t => {
        t.profit = parseFloat((t.profit * scaleFactor).toFixed(2));
    });

    trades.forEach(t => {
         if (t.profit < 2.00) t.profit = 2.00;
         if (t.profit > 8.00) t.profit = 8.00;
         // Recalculate Close Price
         if (t.type === 'buy') {
             t.closePrice = parseFloat((t.openPrice + t.profit).toFixed(2));
         } else {
             t.closePrice = parseFloat((t.openPrice - t.profit).toFixed(2));
         }
    });


    // Add Withdrawal
    const withdrawalDate = new Date(2025, 9, 24, 10, 5, 42);
    trades.push({
        type: 'balance',
        volume: 0,
        symbol: '',
        openPrice: 0,
        closePrice: 0,
        openTimeDate: withdrawalDate,
        closeTimeDate: withdrawalDate,
        profit: -150,
        sl: 0,
        tp: 0,
        commission: 0,
        swap: 0
    });

    // Sort by close time
    trades.sort((a, b) => a.closeTimeDate - b.closeTimeDate);

    // Assign IDs
    let startId = 514343326;
    const finalTrades = trades.map(t => {
        let deal = {
            id: (startId++).toString(),
            symbol: t.symbol,
            type: t.type,
            volume: t.volume,
            openTime: formatDate(t.openTimeDate),
            closeTime: formatDate(t.closeTimeDate),
            openPrice: t.openPrice,
            closePrice: t.closePrice,
            sl: t.sl,
            tp: t.tp,
            commission: t.commission,
            swap: t.swap,
            profit: t.profit
        };
        
        if (t.type === 'balance') {
            deal.symbol = '';
            deal.openPrice = 0;
            deal.closePrice = 0;
            deal.openTime = deal.closeTime;
        }
        
        return deal;
    });

    console.log(JSON.stringify(finalTrades, null, 2));

} catch (error) {
    console.error(error);
}
